### PR TITLE
fix: Prevent a single metric have more than 100 data points

### LINKED ISSUE #311: Bug: Prevent a single metric have more than 100 data points
### Expected Behaviour

Flush metric set when data point is reached to prevent potential data loss, where 100+ data points are added thus being ignored silently by CloudWatch EMF when processed in the background.

[docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html#CloudWatch_Embedded_Metric_Format_Specification_structure_target](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html#CloudWatch_Embedded_Metric_Format_Specification_structure_target)

### Current Behaviour

More than 100 metric data points can be added.

### Code snippet

```csharp
using AWS.Lambda.Powertools.Metrics;

public class Function
{
    [Metrics(CaptureColdStart = true)]
    public async Task<APIGatewayProxyResponse> FunctionHandler(APIGatewayProxyRequest apigwProxyEvent,
        ILambdaContext context)
    {

        Metrics.AddMetric("SuccessfulLocations", 1, MetricUnit.Count);
        Metrics.AddMetric("SuccessfulLocations", 1, MetricUnit.Count);
        // add other 100 values to the same metric
    }
}
```

### Possible Solution

Serialize and flush metric data set when a single metric reaches 100 data points. 
Throw exception when adding a metric with 100 data points.
Dimensions etc should be kept intact.

### Steps to Reproduce

Run the code added in code snippet

### Powertools for AWS Lambda (.NET) version

latest

### AWS Lambda function runtime

dotnet6

### Debugging logs

_No response_

### PR DESCRIPTION
**Pull request by hjgraca**
> Please provide the issue number

Issue number: #311 

## Summary

Serialize and flush metric data set when a single metric reaches 100 data points.
Throw exception when adding a metric with 100 data points.
Dimensions etc should be kept intact.

### Changes

> Please provide a summary of what's being changed

Check if single metrics has more than 100 datapoints and flush. 
Add tests to check first and second metrics flush.

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

Please leave checklist items unchecked if they do not apply to your change.

* [x] [Meets tenets criteria](https://docs.powertools.aws.dev/lambda-dotnet/tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-dotnet/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by hjgraca** [comment]
cc @leandrodamascena

File: libraries/src/AWS.Lambda.Powertools.Metrics/Model/MetricDirective.cs [hunk 1]
```diff
@@ -141,7 +141,13 @@ public void AddMetric(string name, double value, MetricUnit unit, MetricResoluti
         {
             var metric = Metrics.FirstOrDefault(m => m.Name == name);
             if (metric != null)
-                metric.AddValue(value);
+            {
+                if(metric.Values.Count < PowertoolsConfigurations.MaxMetrics)
+                    metric.AddValue(value);
+                else
```
Comment (hjgraca) [inline libraries/src/AWS.Lambda.Powertools.Metrics/Model/MetricDirective.cs:147]: I don't believe we will ever be in this situation (it only accepts one double not an array) but added the throwing exception for consistency with the bellow throw

File: libraries/tests/AWS.Lambda.Powertools.Metrics.Tests/EMFValidationTests.cs [hunk 2]
```diff
@@ -139,20 +139,75 @@ public void When100MetricsAreAdded_FlushAutomatically()
             for (var i = 0; i <= 100; i++)
             {
                 Metrics.AddMetric($"Metric Name {i + 1}", i, MetricUnit.Count);
-            }
 
+                if (i == 100)
```
Comment (hjgraca) [inline libraries/tests/AWS.Lambda.Powertools.Metrics.Tests/EMFValidationTests.cs:143]: I wanted to make sure and test that the first flush would happen and what are the values

**Review by amirkaws — COMMENTED** [review]
Minor changes to use PowertoolsConfigurations.MaxMetrics instead of hardcoded value of 100.

**Review by amirkaws — APPROVED** [review]
LGTM

**Pull request merged**